packages/effect/src/internal/stream.ts

```diff
@@ -4686,6 +4686,39 @@ export const range = (min: number, max: number, chunkSize = DefaultChunkSize): S
     return new StreamImpl(go(min, max, chunkSize))
   })
 
+export const raceAll = <T extends ReadonlyArray<Stream.Stream<any, any, any>>>(
+  ...streams: T
+): Stream.Stream<
+  [T[number]] extends [never] ? never : T[number] extends Stream.Stream<infer A, any, any> ? A : never,
+  [T[number]] extends [never] ? never : T[number] extends Stream.Stream<any, infer E, any> ? E : never,
+  [T[number]] extends [never] ? never : T[number] extends Stream.Stream<any, any, infer R> ? R : never
+> => {
+  let finished = false
```

Comment:
Any state needs to be constructed in the context of the running stream (as they are lazy). So inside the Effect.gen in this case.

Reply:
this is the most important comment, thank you